Unit tests, core bugs

DUznanski opened the pull request:
Tests the core MGL and MGLT functions.  Does not do anything with the actual types yet.

* code is now tested to work with Lua 5.1-5.4 and also LuaJIT
* Includes a Windows batch file to perform the testing

In addition to actually making sure stuff works, fixes several bugs:

* MGL no longer requires special require path handling
* MGLT functions are now properly exposed
* You can now publish your own op changes if you want to (you probably don't want to do that)
* Functions will retry with types before the last as wildcards if it can't find one that matches deeper
* Over-long function calls are now called out as errors instead of silently discarding overflowed arguments.

DUznanski committed:
Begin unit testing process; fix persnickety module loading

- Added luaunit
- Began a testing file and a Windows batch for running tests against 5.1-5.4 and luajit
- Changed module loading system so it doesn't need to have the src folder itself as part of the package path

DUznanski committed:
MGLT testing

Basic testing of the MGLT utilities.
- mglt simply wasn't exposed to the outside world, because `require` adjusts returns to one item.  It's now available as mgl.tools

DUznanski committed:
Add testing for the callback setup

- Make the callback call its own function so I can test it neatly.
- It's one big test because you can't really do them separately.

DUznanski committed:
Test defof/getop/genop

- Wildcards collect fallthroughs at any depth: f(number, table) matches f(*, table) and not f(number, number) but the first previously prevented the second from running.
- overlong function calls are now detected

DUznanski committed:
remember to finish sentences

ImagicTheCat commented:
I'm not currently actively working on the project, but I can give you insights on the pull request.

Issues/notes:
- The pull request do a lot of things, which complicates reviewing it.
- https://github.com/ImagicTheCat/MGL/pull/3/commits/3302e987e715f2601c28f43ae7d823eff88e93c2
  - I don't see the point of the module load fix/trick. I consider Lua modules as an abstraction which associates a module path (not a file path) to the module data. If the use case is about moving the module files, there is `package.path`. If the use case is disambiguation of two modules with the same path, this conflicts with the definition of a module (e.g. two C modules with the same C symbol).
  - I don't want to support Windows as a development platform (no .bat), because of (IMHO) sloppy designs. For example, it seems that there are case insensitive issues for the unit testing.
  - I don't know if it's useful to have automated testing for the core of MGL, but it can be a nice addition. But I don't use unit testing frameworks (unless their advantages are worth the burden), as you can see in `examples/`.
- https://github.com/ImagicTheCat/MGL/pull/3/commits/5305e5e5a00283aa668546b8aac8643a65596dfb
  - I didn't realize that `require` was only made for one return value, interesting, I really expected it to have that behavior.
- https://github.com/ImagicTheCat/MGL/pull/3/commits/a760652c352119efa1bb0dbaa4ed3afa73d46cd9
  - I don't understand that the testing process implies changes of the tested API; `publishOps` has no reasons to be exposed.
- https://github.com/ImagicTheCat/MGL/pull/3/commits/3955ace02056ea77bd1365b38690a919437d31fb
  - I need to think more about the operator branching changes. The goal is to have a simple prototype system to simplify the use of the library, not a complete GLSL syntax. But there is probably room for improvement.

ImagicTheCat commented:
I think the overflow check fix is not sufficient, because we can have an arbitrary number of arguments after that nil (which can be a mistake too). It could be done by comparing the number of arguments passed with the number of expected arguments, but I don't think this is reasonable from the Lua side.

Anyway, after thinking about it I concluded that a dynamic type system deserves its own well thought out library. I will work on that and MGL or other libraries could depend on it.
Such a library may have the following advantages:
- Interoperability: from MGL types to cdata or any third-party types, with any kind of functions.
- Operator/function overloading: multiple signatures, with "abstract" types like the wildcard, but also with a type hierarchy (e.g. for OOP).
- Prevent reinventing the wheel when some automated dynamic type checking is needed.
- Generic functions generated on-the-fly ?

ImagicTheCat commented:
I think I just solved the GLSL constructor pattern issue in my last post without realizing it. The problem with GLSL constructors is the number of possibilities they can have, especially with variable vector size, because we have to generate all of them.

But we can consider Lua as a meta-programming language for itself; Lua code can be generated by Lua code and MGL already does that. We don't have the constraint of static languages, so why not generate things on-the-fly ? Not only this would prevent generating unused functions, but it can also simplify the generation code: the vec/mat constructors can be generated when they are used, in that specific configuration, instead of generating all of the possibilities ahead of time (some JIT vibes in that sentence).

DUznanski committed:
take publishOps back out

now we'll just use defOp directly to show publication working.

Also some more clarity on call mechanics

DUznanski committed:
Hide the testing batch script

Beter idea anyway, you'd need to make your own for however your system holds the various luas

DUznanski commented:
high arity
----

The reasoning behind blocking higher arity function calling here is partly because we already "disable" them silently *anyway*, but we only "allow" it when the arity of the pattern we match against is equal to the highest arity of any pattern we have selected for the function.  Which is to say, in the previously existing system, if we have `f(table)` and `f(number, number)` as allowed patterns, calling `f(table, table)` fails with an "invalid operator prototype" error, but calling `f(number, number, number)` "succeeds" but only passes the first two entries (like Lua, true, but...).  So explicitly disallowing them makes the behavior of MGL's generated dispatch more consistent with itself.  Actually originally that was a thing I ended up with as a workaround - you'll notice I basically pad the arguments with nil to make the generated code easier to generate - but once I realized that it would be more self consistent it felt more like a feature than a kludge or bug.  Also I just added a test to see - adding multiple extra arguments still triggers the error, as intended (which is no surprise: the dispatch function doesn't even actually receive those arguments).

on-the-fly generation and wildcards
----

Part of my plan for the next part (once I deal with some of the above issues) is to do the on-the-fly constructor generation, yes.    It is required that the wildcard fallthrough is there for generation to work - without it, I can't make `vec4(vec2, vec2)` and then `vec4(vec2, number, number)` via on-the-fly generation.  And yeah doing all ahead of time would be problematic: there are 66,546 prototypes for mat4.  I'm also going to want to make a few toys to let that happen "early".

module loading and package paths
----

The thing with the modules is ... if MGL were a library where I would be manually loading individual submodules (like, say, [HUMP](https://github.com/HDictus/hump/tree/temp-master)), it would make a certain level of sense to insist that `package.path` include the interior of MGL's source directory - though libraries I find that are multiple submodules don't do that.  It's not though: the additional files are loaded automatically and "invisibly" by the primary module, and are mainly there as an organization tool for the developers *of MGL*, so building the library so the package path doesn't need to directly include the innards is a reasonable expectation.  This is also the approach taken by [CPML](https://github.com/excessive/cpml/blob/master/init.lua#L35), [STI](https://github.com/karai17/Simple-Tiled-Implementation/blob/master/sti/init.lua#L17), and [Batteries](https://github.com/1bardesign/batteries/blob/master/init.lua#L8), among others.    In fact, most Lua environments (luajit, 5.2, and up) include `?/init.lua` as part of their default `package.path`, which would make renaming the main file to `init.lua` make even more sense (and the other libraries mentioned do this too).

testing script
----

Turns out different people have their various lua binaries set up differently (or don't have them all installed) so including the batch file is not that great an idea even if windows were directly supported.

publishOps
----
Yeah that really should have been taken out once I got the defop code actually tested.

testing framework
----

 * Test **ALL** the things!  If it's not tested we don't know it works.
 * If I were to do this from the beginning without an existing testing framework, the first thing I'd do would be to make something that ends up looking an awful lot like an existing testing framework, because the advantages - automatic collection of testing functions, failure of a single test not interrupting overall testing, prebuilt special assertion types, and others - outweigh the relatively small cost of use.

ImagicTheCat commented:
See https://github.com/ImagicTheCat/MGL/commit/815f11fbbd74aa611847b621049d24ac5716a31b and [xtype](https://github.com/ImagicTheCat/lua-xtype).
